core/common/src/main/java/tachyon/wire/BlockInfo.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.wire;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Objects;
+
+/**
+ * The block descriptor.
+ */
+@NotThreadSafe
+public final class BlockInfo {
+  @JsonProperty("blockId")
+  private long mBlockId;
+  @JsonProperty("length")
+  private long mLength;
+  @JsonProperty("locations")
+  private List<BlockLocation> mLocations = new ArrayList<BlockLocation>();
+
+  /**
+   * Creates a new instance of {@link BlockInfo}.
+   */
+  public BlockInfo() {}
```

Comment:
I agree with your analysis of pros and cons. 

I personally prefer to avoid the Foo(arg1, ..., argn) constructor because they are much harder to maintain. Introducing a new member requires changes to the constructor and all of its invocations or introduces a new constructor; neither of the two options is great.

In contrast to that, the empty constructor + setters paradigm is easier to maintain. After adding a member and its setter, I only need to update invocations of the constructor that require the non-default value to be set. This can save a lot of time if the number of uses of that constructor is large.